SQL types are wrong for shared primary keys

This module demonstrates the problem.

```haskell
{-# LANGUAGE TypeApplications, DeriveGeneric #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE StandaloneDeriving #-}

module SharedPrimaryKeyTest where

import TemplateTestImports

import Database.Persist
import Database.Persist.Sql
import Database.Persist.Sql.Util
import Database.Persist.TH

share [ mkPersist sqlSettings ] [persistLowerCase|

User
    name    String

Profile
    Id      UserId
    email   String

|]

specs :: Spec
specs = describe "Shared Primary Keys" $ do
```

As of now, this doesn't even compile.

```
/home/matt/Projects/persistent/persistent-template/test/SharedPrimaryKeyTest.hs:23:51: error:
    • Not in scope: type constructor or class ‘UserId’
    • In the quasi-quotation:
        [persistLowerCase|
            
User        
    name    String
            
Profile     
    Id      UserId
    email   String
            
|]          
   |        
23 | share [ mkPersist sqlSettings ] [persistLowerCase|
   |                                                   ^...
   
```

Switching `Profile` to using a `userId UserId; Primary userId` works. But that's pretty weird, I'd except `UserId` to have the same resolution with `Id` as it would for `userId`.

You can get Clever by doing a `(Key User)` instead. Then the `PersistFieldSql` is right, but the actual `entityId` on the field definition is wrong.

--- a/.github/workflows/haskell.yml
+++ b/.github/workflows/haskell.yml
@@ -63,7 +63,8 @@ jobs:
             ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
           key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
           restore-keys: |
-            ${{ runner.os }}-${{ matrix.ghc }}-
+            ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
+            #  ${{ runner.os }}-${{ matrix.ghc }}-
       - run: cabal v2-build all --disable-optimization --only-dependencies $CONFIG
       - run: cabal v2-build all --disable-optimization $CONFIG
       - run: cabal v2-test all --disable-optimization $CONFIG

--- a/.github/workflows/haskell.yml
+++ b/.github/workflows/haskell.yml
@@ -61,7 +61,6 @@ jobs:
         with:
           path: |
             ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
-            dist-newstyle
           key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
           restore-keys: |
             ${{ runner.os }}-${{ matrix.ghc }}-

--- a/persistent-template/ChangeLog.md
+++ b/persistent-template/ChangeLog.md
@@ -1,5 +1,12 @@
 ## Unreleased changes
 
+## 2.9.1.0
+
+* [#1145](https://github.com/yesodweb/persistent/pull/1148)
+    * Fix a bug where the `SqlType` for a shared primary key was being
+      incorrectly set to `SqlString` instead of whatever the target primary key
+      sql type was.
+
 ## 2.9
 
 * Always use the "stock" strategy when deriving Show/Read for keys [#1106](https://github.com/yesodweb/persistent/pull/1106)

--- a/persistent-template/Database/Persist/TH.hs
+++ b/persistent-template/Database/Persist/TH.hs
@@ -1677,7 +1677,7 @@ liftAndFixKeys entityMap EntityDef{..} =
     [|EntityDef
         entityHaskell
         entityDB
-        entityId
+        $(liftAndFixKey entityMap entityId)
         entityAttrs
         $(ListE <$> mapM (liftAndFixKey entityMap) entityFields)
         entityUniques

--- a/persistent-template/persistent-template.cabal
+++ b/persistent-template/persistent-template.cabal
@@ -1,5 +1,5 @@
 name:            persistent-template
-version:         2.9
+version:         2.9.1.0
 license:         MIT
 license-file:    LICENSE
 author:          Michael Snoyman <michael@snoyman.com>
@@ -40,6 +40,7 @@ test-suite test
     other-modules:   
         TemplateTestImports
       , SharedPrimaryKeyTest
+      , SharedPrimaryKeyTestImported
                     
     ghc-options:     -Wall
 

--- a/persistent-template/persistent-template.cabal
+++ b/persistent-template/persistent-template.cabal
@@ -37,7 +37,10 @@ test-suite test
     type:            exitcode-stdio-1.0
     main-is:         main.hs
     hs-source-dirs:  test
-    other-modules:   TemplateTestImports
+    other-modules:   
+        TemplateTestImports
+      , SharedPrimaryKeyTest
+                    
     ghc-options:     -Wall
 
     build-depends:   base                     >= 4.10 && < 5

--- a/persistent-template/test/SharedPrimaryKeyTest.hs
+++ b/persistent-template/test/SharedPrimaryKeyTest.hs
@@ -27,7 +27,7 @@ share [ mkPersist sqlSettings ] [persistLowerCase|
 User
     name    String
 
--- TODO: uncomment this out.
+-- TODO: uncomment this out https://github.com/yesodweb/persistent/issues/1149
 -- Profile
 --     Id      UserId
 --     email   String

--- a/persistent-template/test/SharedPrimaryKeyTest.hs
+++ b/persistent-template/test/SharedPrimaryKeyTest.hs
@@ -1,0 +1,58 @@
+{-# LANGUAGE TypeApplications, DeriveGeneric #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module SharedPrimaryKeyTest where
+
+import TemplateTestImports
+
+import Data.Proxy
+import Test.Hspec
+import Database.Persist
+import Database.Persist.Sql
+import Database.Persist.Sql.Util
+import Database.Persist.TH
+
+share [ mkPersist sqlSettings ] [persistLowerCase|
+
+User
+    name    String
+
+-- TODO: uncomment this out.
+-- Profile
+--     Id      UserId
+--     email   String
+
+Profile
+    Id      (Key User)
+    email   String
+
+|]
+
+spec :: Spec
+spec = describe "Shared Primary Keys" $ do
+
+    describe "PersistFieldSql" $ do
+        it "should match underlying key" $ do
+            sqlType (Proxy @UserId)
+                `shouldBe`
+                    sqlType (Proxy @ProfileId)
+
+    describe "entityId FieldDef" $ do
+        it "should match underlying primary key" $ do
+            let userIdDef =
+                    fieldSqlType $ entityId $ entityDef $ Proxy @User
+                profileIdDef =
+                    fieldSqlType $ entityId $ entityDef $ Proxy @Profile
+            profileIdDef
+                `shouldBe`
+                    userIdDef

--- a/persistent-template/test/SharedPrimaryKeyTestImported.hs
+++ b/persistent-template/test/SharedPrimaryKeyTestImported.hs
@@ -11,7 +11,7 @@
 {-# LANGUAGE DerivingStrategies #-}
 {-# LANGUAGE StandaloneDeriving #-}
 
-module SharedPrimaryKeyTest where
+module SharedPrimaryKeyTestImported where
 
 import TemplateTestImports
 
@@ -22,24 +22,21 @@ import Database.Persist.Sql
 import Database.Persist.Sql.Util
 import Database.Persist.TH
 
+import SharedPrimaryKeyTest (User, UserId)
+
 share [ mkPersist sqlSettings ] [persistLowerCase|
 
-User
-    name    String
-
--- TODO: uncomment this out.
--- Profile
---     Id      UserId
---     email   String
-
 Profile
-    Id      (Key User)
+    Id      UserId
     email   String
 
 |]
 
+-- This test is very similar to the one in SharedPrimaryKeyTest, but it is
+-- able to use 'UserId' directly, since the type is imported from another
+-- module.
 spec :: Spec
-spec = describe "Shared Primary Keys" $ do
+spec = describe "Shared Primary Keys Imported" $ do
 
     describe "PersistFieldSql" $ do
         it "should match underlying key" $ do

--- a/persistent-template/test/main.hs
+++ b/persistent-template/test/main.hs
@@ -41,6 +41,7 @@ import Database.Persist.Sql.Util
 import Database.Persist.TH
 import TemplateTestImports
 
+import qualified SharedPrimaryKeyTest
 
 share [mkPersist sqlSettings { mpsGeneric = False, mpsDeriveInstances = [''Generic] }, mkDeleteCascade sqlSettings { mpsGeneric = False }] [persistUpperCase|
 Person json
@@ -106,6 +107,7 @@ instance Arbitrary Address where
 
 main :: IO ()
 main = hspec $ do
+    SharedPrimaryKeyTest.spec
     describe "hasNaturalKey" $ do
         let subject :: PersistEntity a => Proxy a -> Bool
             subject p = hasNaturalKey (entityDef p)

--- a/persistent-template/test/main.hs
+++ b/persistent-template/test/main.hs
@@ -42,6 +42,7 @@ import Database.Persist.TH
 import TemplateTestImports
 
 import qualified SharedPrimaryKeyTest
+import qualified SharedPrimaryKeyTestImported
 
 share [mkPersist sqlSettings { mpsGeneric = False, mpsDeriveInstances = [''Generic] }, mkDeleteCascade sqlSettings { mpsGeneric = False }] [persistUpperCase|
 Person json
@@ -108,6 +109,7 @@ instance Arbitrary Address where
 main :: IO ()
 main = hspec $ do
     SharedPrimaryKeyTest.spec
+    SharedPrimaryKeyTestImported.spec
     describe "hasNaturalKey" $ do
         let subject :: PersistEntity a => Proxy a -> Bool
             subject p = hasNaturalKey (entityDef p)

